test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs

```diff
@@ -0,0 +1,214 @@
+            this.ResultType = Preconditions.CheckNonWhiteSpace(resultType, nameof(resultType));
+            this.NetworkStatusTimeline = Preconditions.CheckNotNull(networkStatusTimeline, nameof(networkStatusTimeline));
+        }
+
+        internal string ActualSource { get; }
+
+        internal ITestResultCollection<TestOperationResult> ActualTestResults { get; }
+
+        internal string ExpectedSource { get; }
+
+        internal ITestResultCollection<TestOperationResult> ExpectedTestResults { get; }
+
+        internal string ResultType { get; }
+
+        internal ITestResultComparer<TestOperationResult> TestResultComparer { get; }
+
+        internal Option<NetworkStatusTimeline> NetworkStatusTimeline { get; }
+
+        public async Task<ITestResultReport> CreateReportAsync()
+        {
+            Logger.LogInformation($"Start to generate report by {nameof(DirectMethodReportGenerator)} for Sources [{this.ExpectedSource}] and [{this.ActualSource}]");
+
+            ulong networkOnSuccess = 0;
+            ulong networkOffSuccess = 0;
+            ulong networkOnToleratedSuccess = 0;
+            ulong networkOffToleratedSuccess = 0;
+            ulong networkOnFailure = 0;
+            ulong networkOffFailure = 0;
+            ulong mismatchSuccess = 0;
+            ulong mismatchFailure = 0;
+
+            bool hasExpectedResult = await this.ExpectedTestResults.MoveNextAsync();
+            bool hasActualResult = await this.ActualTestResults.MoveNextAsync();
+
+            while (hasExpectedResult && hasActualResult)
+            {
+                this.ValidateDataSource(this.ExpectedTestResults.Current, this.ExpectedSource);
+                this.ValidateDataSource(this.ActualTestResults.Current, this.ActualSource);
+
+                (NetworkControllerStatus networkControllerStatus, bool isWithinTolerancePeriod) =
+                        this.NetworkStatusTimeline.Expect<InvalidOperationException>(
+                            () => throw new InvalidOperationException("Results are present, but NetworkStatusTimeline is empty."))
+                        .GetNetworkControllerStatusAndWithinToleranceAt(this.ExpectedTestResults.Current.CreatedAt);
+
+                if (this.TestResultComparer.Matches(this.ExpectedTestResults.Current, this.ActualTestResults.Current))
+                {
+                    // Found same message in both stores.
+                    if (NetworkControllerStatus.Disabled.Equals(networkControllerStatus))
+                    {
+                        // If network on at time, succeed
+                        networkOnSuccess++;
+                    }
+                    else if (NetworkControllerStatus.Enabled.Equals(networkControllerStatus))
+                    {
+                        // If network off at time, fail unless in tolerance period
+                        if (isWithinTolerancePeriod)
+                        {
+                            networkOffToleratedSuccess++;
```

Comment:
I don't think so, since, in a way, the whole report is a log of what happened. But this is just report generation based on events that have already happened, so I don't think logging is as useful here